### ISSUE TITLE
Out of memory while synching blockchain

### ISSUE THREAD

**Issue by patricklodder** [dogecoin/dogecoin#714]
I got this report on freenode today, logging here for info / discussion.

-----

The user is synching dogecoind on a 32-bit linux vps with 512M RAM and gets:

console:
```
dogecoind: main.cpp:1890: bool ConnectBlock(CBlock&, CValidationState&, CBlockIndex*, CCoinsViewCache&, bool): 
  Assertion `hashPrevBlock == view.GetBestBlock()' failed.
```

and in debug.log:
```
2014-09-22 11:00:55 ProcessBlock: ACCEPTED
2014-09-22 11:00:55 GetNextWorkRequired() : RETARGET; target: 60, actual: 4, modulated: 53, before: 1b4fed98, after: 1b469a68
2014-09-22 11:00:56 LoadExternalBlockFile : Deserialize or I/O error - std::bad_alloc
GetNextWorkRequired() : RETARGET; target: 60, actual: 144, modulated: 70, before: 1b469a68, after: 1b525ece
```

Reported to happen at different blocks, this log is for block [225818](https://chain.so/block/DOGE/0f5f8806a71c68d064685f5e972c92d9e96d8791a0fbc271be59eabc40de221c) ([json](https://chain.so/api/v2/block/DOGE/0f5f8806a71c68d064685f5e972c92d9e96d8791a0fbc271be59eabc40de221c))

User reports this happens with both network sync and bootstrap.dat.

-----

I'll set up a 32-bit env and run some checks with valgrind later, see if there's something obviously wrong, or that it's just the fact that there's a [70kb tx inside the block in question, with 476 inputs](https://chain.so/tx/DOGE/bf44b8ee6183463c349f87d8bf0c097ac0d1b76fbbaa3fbc6ffda3049159c0e7).

**Comment by prestotron** [comment]
I just documented the Unix build instructions with directions for adding swap on lower RAM devices to fix this issue.

It hasn't been merged as of this post, but it might soon. Just hope :)

**Comment by rnicoll** [comment]
1.10 reduces memory footprint, but honestly I think 512mb is optimistic given the size of the dataset. Any objections to closing this?

**Comment by patricklodder** [comment]
Can we let it open until I have tested a 1.10 instance with all our crap in? :)

**Comment by patricklodder** [comment]
> 1.10 reduces memory footprint

For me, 1.10.0 takes about double the memory 1.8.3 takes. 
@rnicoll: thoughts on why, if we were expecting the opposite?

**Comment by patricklodder** [comment]
I've done some selective reverts of PRs that I thought could have caused this, and the winner is: #1206 

With this patch dogecoind takes about ~950MB resident memory, without it ~380MB.

**Comment by rnicoll** [comment]
Ah, yes, of course we're adding a whole second header and additional merkle branches to the index. Might be an idea to load it with the AuxPoW header as well, then drop it after validation, not sure.

**Comment by patricklodder** [comment]
fixed with #1338 as best we can for now.
